src/main/kotlin/org/jetbrains/research/testspark/helpers/psi/kotlin/KotlinPsiMethodWrapper.kt

```diff
@@ -0,0 +1,108 @@
+
+    override val methodDescriptor: String
+        get() = psiFunction.run {
+            val parameterTypes = valueParameters.joinToString("") { generateFieldType(it.typeReference) }
+            val returnType = generateReturnDescriptor(psiFunction)
+            return "$name($parameterTypes)$returnType"
+        }
+
+    override val signature: String
+        get() = psiFunction.run {
+            val bodyStart = bodyExpression?.startOffsetInParent ?: textLength
+            text.substring(0, bodyStart).replace('\n', ' ').trim()
+        }
+
+    val parameterList = psiFunction.valueParameterList
+
+    val isPrimaryConstructor: Boolean = psiFunction is KtPrimaryConstructor
+
+    val isSecondaryConstructor: Boolean = psiFunction is KtSecondaryConstructor
+
+    val isTopLevelFunction: Boolean = psiFunction.containingClassOrObject == null
+
+    val isDefaultMethod: Boolean = psiFunction.run {
+        val containingClass = PsiTreeUtil.getParentOfType(this, KtClassOrObject::class.java)
+        val containingInterface = containingClass?.isInterfaceClass()
+        name != "<init>" && bodyExpression != null && containingInterface == true
+    }
+
+    override fun containsLine(lineNumber: Int): Boolean {
+        val psiFile = psiFunction.containingFile
+        val document = PsiDocumentManager.getInstance(psiFile.project).getDocument(psiFile) ?: return false
+        val textRange = psiFunction.textRange
+        val startLine = document.getLineNumber(textRange.startOffset) + 1
+        val endLine = document.getLineNumber(textRange.endOffset) + 1
+        return lineNumber in startLine..endLine
+    }
+
+    /**
+     * Generates the return descriptor for a method.
+     *
+     * @param psiFunction the function
+     * @return the return descriptor
+     */
+    private fun generateReturnDescriptor(psiFunction: KtFunction): String {
+        val returnType = psiFunction.typeReference?.text ?: "Unit"
+        return generateFieldType(returnType)
+    }
+
+    /**
+     * Generates the field descriptor for a type.
+     *
+     * @param typeReference the type reference to generate the descriptor for
+     * @return the field descriptor
+     */
+    private fun generateFieldType(typeReference: KtTypeReference?): String {
+        val type = typeReference?.text ?: "Unit"
+        return generateFieldType(type)
+    }
+
+    /**
+     * Generates the field descriptor for a type.
+     *
+     * @param type the type to generate the descriptor for
+     * @return the field descriptor
+     */
+    private fun generateFieldType(type: String): String {
+        return when (type) {
+            "Int" -> "I"
+            "Long" -> "J"
+            "Float" -> "F"
+            "Double" -> "D"
+            "Boolean" -> "Z"
+            "Byte" -> "B"
+            "Char" -> "C"
+            "Short" -> "S"
+            "Unit" -> "V"
+            else -> "L${type.replace('.', '/')};"
+        }
+    }
```

Comment:
Where does this mapping of types and their abbreviations come from? Could place a link to the source in the Javadoc as well?

Reply:
https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.3